### PR TITLE
Regenerated rest api client and Refactored workspace connection parameters

### PR DESCRIPTION
**Pull request by vxfield**
# DRAFT

## Main changes:
1. Regenerated the rest api client based on latest swagger
2. Refactored the parameters to connect to a workspace into its own class
3. Gathered common constants into a new _constants.py file
4. Cleaned up code/logic of the _authentication module
5. Cleaned up and fixed some bugs in the unit test framework
6. Refactored, cleaned up, and added more tests for the workspace connection parameters and authentication

## To-do:
1. Fix remaining issues with unit test playback
2. Finish validating the app_key/connection_string client support (pending dogfood/canary deployment)
3. Re-record unit tests

### CODE REVIEW

File: azure-quantum/azure/quantum/_authentication/_default.py [hunk 1]
```diff
@@ -54,162 +52,86 @@ class _DefaultAzureCredential(_ChainedTokenCredential):
...
+            credentials.append(VisualStudioCodeCredential(authority=self.authority, tenant_id=self.tenant_id))
+            credentials.append(AzureCliCredential(tenant_id=self.tenant_id))
+            credentials.append(AzurePowerShellCredential(tenant_id=self.tenant_id))
+            credentials.append(InteractiveBrowserCredential(authority=self.authority, tenant_id=self.tenant_id))
+            if self.client_id:
+                credentials.append(DeviceCodeCredential(authority=self.authority, client_id=self.client_id, tenant_id=self.tenant_id))
         self.credentials = credentials
 
     def get_token(self, *scopes, **kwargs):
-        # type: (*str, **Any) -> AccessToken
```
Comment (xinyi-joffre) [inline azure-quantum/azure/quantum/_authentication/_default.py:141]: It looks like we removed the type hinting for the return type. Is this expected?

Comment (xinyi-joffre) [reply azure-quantum/azure/quantum/_authentication/_default.py:141]: @kikomiss, do you know if the type hints are needed for doc generation?

File: azure-quantum/azure/quantum/workspace.py [hunk 2]
```diff
@@ -131,111 +114,97 @@ def __init__(
         resource_id: Optional[str] = None,
         location: Optional[str] = None,
         credential: Optional[object] = None,
-        user_agent: Optional[str] = None,
+        connection_string: Optional[str] = None,
```
Comment (xinyi-joffre) [inline azure-quantum/azure/quantum/workspace.py:117]: Did this change from the spec review, Yi? @04diiguyi , I thought we had to do something like Workspace().from_connection_string("<connection_string>"): 
https://azure.github.io/azure-sdk/python_design.html#azure-sdk-api-design
image

Comment (04diiguyi) [reply azure-quantum/azure/quantum/workspace.py:117]: Yes. @vxfield , let's sync up on this. Please follow the python sdk rule we have in this PR. https://github.com/microsoft/azure-quantum-python/pull/556/commits/3cf130002f2d5058c0715a7158bc551a667719f1

**Pull request closed without merging**